media/mp4.ksy

```diff
@@ -0,0 +1,408 @@
+meta:
+  id: mp4
+  file-extension: mp4
+  endian: be
+  license: CC0-1.0
+  encoding: ascii
+  
+#
+# The container is a tree-like structure of boxes.
+# There are various types of boxes. Boxes of some types can contain nested boxes.
+# You can find some interesting information in those boxes:
+#
+# -> moov -> mvhd -> duration / timescale    :: this is how you calculate media length in seconds
+# -> moov -> trak                            :: those are tracks in the container (video/audio/etc.)
+# trak -> tkhd -> @width x @height           :: resolution of the video track
+# trak -> mdia -> minf -> stbl -> stsd.entries[0].type :: fourcc of codec used for the track
+# trak -> mdia -> mdhd -> langCode           :: language code of the track
+#
+# trak -> mdia -> stb -> stts                :: table to get sample index for given time
+# trak -> mdia -> stb -> stsc                :: table to get chunk index for given time
+# trak -> mdia -> stb -> stco                :: table to get offset in file for given chunk index
+#
+# See also:
+# * http://mp4parser.com/
+# * https://archive.codeplex.com/?p=mp4explorer
+# * https://github.com/axiomatic-systems/Bento4
+
+seq:
+  - id: boxes
+    type: box
+    repeat: eos
+
+types:
+
+  boxes:
+    seq:
+      - id: boxes
+        type: box
+        repeat: eos
+
+  box:
+    seq:
+      - id: size
+        type: u4
+      - id: type
+        type: str
+        size: 4
+      - id: largesize
+        type: u8
+        if: size == 1
+      - id: body
+        type:
+          switch-on: type
+          cases:
+            '"moov"': boxes
+            '"trak"': boxes
+            '"edts"': boxes
+            '"mdia"': boxes
+            '"stbl"': boxes
+            '"minf"': boxes
+            '"dinf"': boxes
+            '"ftyp"': box_body_ftyp
+            '"hdlr"': box_body_hdlr
+            '"mvhd"': box_body_mvhd
+            '"vmhd"': box_body_vmhd
+            '"mdhd"': box_body_mdhd
+            '"tkhd"': box_body_tkhd
+            '"elst"': box_body_elst
+            '"stsc"': box_body_stsc
+            '"stsd"': box_body_stsd
+            '"stco"': box_body_stco
+            '"stts"': box_body_stts
+            '"co64"': box_body_co64
+            _: box_body_other
+        size: '(size == 1) ? size - 12 : size - 8'
+
+  # -----------
+  # -- FTYP  --
+  # -----------
+  box_body_ftyp:
+    seq:
+      - id: major_brand
+        type: str
+        size: 4
+      - id: minor_version
+        type: str
+        size: 4
+      - id: compatible_brands
+        type: str
+        size: 4
+        repeat: eos
+#        size-eos: true
+
+  # -----------
+  # -- MVHD  --
+  # -----------
+  box_body_mvhd:
+    seq:
+      - id: version
+        type: u4
+
+      - id: creation_time
+        type:
+          switch-on: version
+          cases:
```

Comment:
I see pretty many records like tyis one, shoult tyis selector be encapsulated into a type?